Software/Development/SemesterProject/App/referenceHandler.c

```diff
@@ -41,66 +109,67 @@ void decideReferenceType(void)
     }break;
     case cruiseControl:
     {
-        if (referenceTypeIsChanged())   referenceType = torqueControl;
+        if (referenceTypeIsChanged())
+        {
+            referenceType = torqueControl;
+            restartSpeedReference();
+        }
+        setCruiseControlLED(ON);
     }break;
     }
 }
 
 void calculateReference(void)
 {
-    GUISignals = getGUISignals();
     calculateTorqueReference();
     calculateSpeedReference();
 }
 
 void calculateTorqueReference(void)
 {
-    float torqueReferenceBeforeLimit, torqueReferenceBeforeSaturation;
+    float torqueReferenceBeforeLimit = 0, torqueReferenceBeforeSaturation = 0;
 
-    decideReferenceSource();
     torqueReferenceBeforeLimit = getReferenceValue();
     torqueReferenceBeforeSaturation = calculateLimitedTorqueReference(torqueReferenceBeforeLimit);
     torqueReference = calculateSaturatedTorque(torqueReferenceBeforeSaturation);
 }
 
-void decideReferenceSource(void)
+int referenceSourceIsChanged(void)
 {
-    if (referenceSourceIsChanged())
+    if (GUISignals.ReferenceSourcePushbutton || referenceSourceHasBeenPressed())
     {
-        referenceSource++;
-        if (referenceSource == lastIndexRS) referenceSource = interfacePCB;
+        torqueReference = 0;
+        speedReference = 0;
+        if ((readSystemState() == RUNNING)) referenceSourceChanged = 1; //If the referenceSOurce is changed while in running, qualify the change so the system know it must go back to STANDBY
+        return 1;
     }
-}
-
-int referenceSourceIsChanged(void)
-{
-    return (GUISignals.ReferenceSourcePushbutton);// || torqueReferenceHasBeenPressed()) DO NOT FORGET TO ADD THIS FUNCTION WHEN PUSHBUTTON HANDLER IS MERGED
+    else    return 0;
 }
 
 int referenceTypeIsChanged(void)
 {
-    return (GUISignals.ReferenceTypePushbutton); // || cruiseControlHasBeenPressed()DO NOT FORGET TO ADD THIS FUNCTION WHEN PUSHBUTTON HANDLER IS MERGED
+    return (GUISignals.ReferenceTypePushbutton || referenceTypeHasBeenPressed());
 }
 
 float getReferenceValue(void)
 {
-    float analogValueTorque, torqueReferenceBeforeLimit;
+    float analogValueTorque = 0, torqueReferenceBeforeLimit = 0;
     switch (referenceSource)
     {
     case interfacePCB:
     {
         analogValueTorque = getTorqueReferenceSliderMeasurement();
         torqueReferenceBeforeLimit = CalculateTorqueRefFromAnalog(analogValueTorque);
     }break;
-    case GUI:
-    {
-        torqueReferenceBeforeLimit = GUISignals.TorqueReference;
-    }break;
     case goKart:
     {
         analogValueTorque = getTorqueReferencePedalMeasurement();
         torqueReferenceBeforeLimit = CalculateTorqueRefFromAnalog(analogValueTorque);
     }break;
+    case GUI:
+    {
+        torqueReferenceBeforeLimit = GUISignals.TorqueReference;
+    }break;
     }
     return torqueReferenceBeforeLimit;
 }
```

Comment:
Maybe add comments on the idea behind limiting the reference changes.

Reply:
Done-ish in header¿

Reply:
Awsome¿?¿